src/web_peer_connection.cpp

```diff
@@ -649,14 +658,39 @@ void web_peer_connection::handle_redirect(int const bytes_left)
 		if (web->have_files.get_bit(file_index) == false)
 		{
 			web->have_files.set_bit(file_index);
+			torrent_info const& info = t->torrent_file();
+
+			// let's try to predict file location names.
+			// we should try to do it cause original request will be dropped
+			// and further requests will use exclusive range pick logic
+			// which exclude edge pieces file locations.
+
+			std::string original_path = escape_file_path(info.orig_files(), file_index);
+			std::size_t redirect_path_idx = redirect_path.rfind(original_path);
+			if (redirect_path_idx != std::string::npos) // we can predict new location
+			{
+				std::string redirect_path_prefix = redirect_path.substr(0, redirect_path_idx);
+				file_storage const& fs = info.files();
+				for (file_index_t fi(0); fi < fs.end_file(); ++fi)
+				{
+					if (info.orig_files().pad_file_at(fi))
+						continue;
+					if (web->redirects.find(fi) != web->redirects.end())
+						continue;
+					web->have_files.set_bit(fi);
```

Comment:
that case is supposed to be handled by storing partial pieces in the ``web_seed_entry``. I suppose it's a problem when you combine that with redirects, since you have multiple different ``web_seed_entry`` objects. I would prefer to extend the existing solution than to do this though. It seems a bit hacky, and doesn't solve the general problem (afaict).